Userland/Libraries/LibSQL/TreeNode.cpp

```diff
@@ -0,0 +1,284 @@
+    for (u32 i = 0; i < nodes; i++)
+    {
+        u32 left;
+        at_offset = deserialize_from(buffer, at_offset, left);
+        m_entries.append(Key(m_tree->index_def(), buffer, at_offset));
+        if (left) {
+            m_down_pointers.append(left);
+            m_down_nodes.append(nullptr);
+            VERIFY(size() == m_down_pointers.size());
+        }
+    }
+    u32 right;
+    deserialize_from(buffer, at_offset, right);
+    if (right) {
+        m_down_pointers.append(right);
+        m_down_nodes.append(nullptr);
+        VERIFY((size()  + 1) == m_down_pointers.size());
+    }
+}
+
+bool TreeNode::insert(Key const& key, TreeNode *right)
+{
+    dbgln("[{}] insert({}, {})",
+        m_pointer, (String) key, ((right) ? right->m_pointer : -1));
+    if (!is_leaf()) {
+        dbgln("Node is not a leaf. Find leaf to insert into");
+        return node_for(key)->insert(key);
+    }
+    VERIFY(is_leaf());
+    if (!m_tree->duplicates_allowed()) {
+        for (auto& entry : m_entries) {
+            if (key == entry) {
+                dbgln("BTree does not allow duplicates and key {} already exists");
+                return false;
+            }
+        }
+    }
+
+    dbgln("Inserting in this node");
+    if (m_entries.size() == (size_t)max_entries()) {
+        split_and_insert(key, right);
+        dump();
+        return true;
+    }
+    VERIFY(m_entries.size() < max_entries());
+    just_insert(key, right);
+    dump();
+    return true;
+}
+
+NonnullRefPtr<Key> TreeNode::operator[](size_t ix) const
+{
+    VERIFY(ix < size());
+    return m_entries[ix];
+}
+
+u32 TreeNode::down_pointer(size_t ix) const
+{
+    VERIFY(ix < m_down_pointers.size());
+    return m_down_pointers[ix];
+}
+
+RefPtr<TreeNode> TreeNode::down_node(size_t ix) const
+{
+    VERIFY(ix < m_down_pointers.size());
+    auto ptr = m_down_pointers[ix];
+    auto ret = m_down_nodes[ix];
+    if (!ret) {
+        VERIFY(ptr != 0);
+        auto buffer = tree()->read_block(ptr);
+        size_t offset;
+        ret = adopt_ref(*new TreeNode(tree(), this, ptr, buffer, offset));
+    } else {
+        VERIFY(ret->m_pointer == ptr);
+    }
+    return ret;
+}
+
+RefPtr<TreeNode> TreeNode::node_for(Key const& key)
+{
+    dbgln("[{}] node_for(Key {})", m_pointer, (String) key);
+    dump();
+    if (is_leaf()) {
+        return this;
+    }
+    for (size_t ix = 0; ix < size(); ix++) {
+        if (key < m_entries[ix]) {
+            dbgln("[{}] {} < {} ({} -> {})",
+                m_pointer, (String) key, (String) m_entries[ix],
+                ix, m_down_nodes[ix]->pointer());
+            return down_node(ix)->node_for(key);
+        }
+    }
+    dbgln("[{}] {} >= {} ({} -> {})",
+        m_pointer, (String) key, (String) m_entries[size() - 1],
+        size(), m_down_nodes[size()]->pointer());
+    return down_node(size())->node_for(key);
+}
+
+Optional<u32> TreeNode::get(Key& key)
+{
+    dbgln("[{}] find({})",
+        m_pointer, (String) key);
```

Comment:
is there not a better way than to c-cast it?

Reply:
Sure. I will create a to_string() method and have the (String) operator use that.

Reply:
You don't need to override the operator nor use a c-cast here if you make a to_string() function, just call it directly